python/cudf/cudf/core/buffer/cow_buffer.py

```diff
@@ -0,0 +1,151 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
+import copy
+from collections import defaultdict
+from typing import Any, DefaultDict, Tuple, Type, TypeVar
+from weakref import WeakSet
+
+import rmm
+
+from cudf.core.buffer.buffer import Buffer, cuda_array_interface_wrapper
+
+T = TypeVar("T", bound="CopyOnWriteBuffer")
+
+
+class CopyOnWriteBuffer(Buffer):
+    """A Buffer represents device memory.
+
+    Use the factory function `as_buffer` to create a Buffer instance.
+    """
+
+    # This dict keeps track of all instances that have the same `ptr`
+    # and `size` attributes.  Each key of the dict is a `(ptr, size)`
+    # tuple and the corresponding value is a set of weak references to
+    # instances with that `ptr` and `size`.
+    _instances: DefaultDict[Tuple, WeakSet] = defaultdict(WeakSet)
+
+    # TODO: This is synonymous to SpillableBuffer._exposed attribute
+    # and has to be merged.
+    _zero_copied: bool
+
+    def _finalize_init(self):
+        # the last step in initializing a `CopyOnWriteBuffer`
+        # is to track it in `_instances`:
+        key = (self.ptr, self.size)
+        self.__class__._instances[key].add(self)
+        self._zero_copied = False
+
+    @classmethod
+    def _from_device_memory(cls: Type[T], data: Any) -> T:
+        """Create a Buffer from an object exposing `__cuda_array_interface__`.
+
+        No data is being copied.
+
+        Parameters
+        ----------
+        data : device-buffer-like
+            An object implementing the CUDA Array Interface.
+
+        Returns
+        -------
+        Buffer
+            Buffer representing the same device memory as `data`
+        """
+
+        # Bypass `__init__` and initialize attributes manually
+        ret = super()._from_device_memory(data)
+        ret._finalize_init()
+        return ret
+
+    @classmethod
+    def _from_host_memory(cls: Type[T], data: Any) -> T:
+        ret = super()._from_host_memory(data)
+        ret._finalize_init()
+        return ret
+
+    def _shallow_copied(self):
+        """
+        Return `True` if shallow copies of `self` exist.
+        """
+        return len(self.__class__._instances[(self.ptr, self.size)]) > 1
+
+    def copy(self, deep: bool = True):
+        """
+        Return a copy of Buffer.
+
+        Parameters
+        ----------
+        deep : bool, default True
+            If True, returns a deep-copy of the underlying Buffer data.
+            If False, returns a shallow-copy of the Buffer pointing to
+            the same underlying data.
+
+        Returns
+        -------
+        Buffer
+        """
+        if not deep:
+            copied_buf = CopyOnWriteBuffer.__new__(CopyOnWriteBuffer)
+            copied_buf._ptr = self._ptr
+            copied_buf._size = self._size
+            copied_buf._owner = self._owner
+            copied_buf._finalize_init()
```

Comment:
Why are we not just implementing `CopyOnWriterBuffer.__init__` ? Do you need finer-grained control over what is going on?